bitsandbytes/backends/cuda.py

```diff
@@ -0,0 +1,456 @@
+import ctypes as ct
+from typing import Optional, Tuple
+
+import torch
+
+from bitsandbytes.cextension import lib
+from bitsandbytes.functional import (
+    CUBLAS_Context,
+    coo_zeros,
+    dequantize_blockwise,
+    dtype2bytes,
+    get_4bit_type,
+    get_colrow_absmax,
+    get_ptr,
+    get_transform_buffer,
+    is_on_gpu,
+    post_call,
+    pre_call,
+    prod,
+    quantize_blockwise,
```

Comment:
Since some of these in `bitsandbytes.functional` are dependent on CUDA for implementation, I'm thinking that maybe they should be moved in here to try and isolate everything CUDA-specific and have `bitsandbytes.functional` be device-agnostic. That can happen in future PRs, though. So far this is great!

Reply:
Yes, to keep minimal changes here for `bitsandbytes.functional` we import these implementations. Will follow in future PR to make `bitsandbytes.functional` device-agnostic. Thanks for your advice!